File: python_unity_shared_memory/test/test_sm.py
from python_unity_shared_memory import  SharedMemory, delete_files
import os
import pytest

def test_create():
    delete_files("test")
    sm = SharedMemory("test", 100, True, 30)
    sm.close()
    assert os.path.exists(sm._hook.file_path)
    delete_files("test")
    assert not os.path.exists(sm._hook.file_path)
    
def test_read_write():
    delete_files("test")
    sm = SharedMemory("test", 100, True, 30)

    assert not sm.read_bool(0)[0]
    sm.write_bool(0, True)
    assert sm.read_bool(0)[0]

    assert sm.read_int32(3)[0] == 0
    sm.write_int32(3, 42)
    assert sm.read_int32(3)[0] == 42

    assert sm.read_float32(10)[0] == 0
    sm.write_float32(10, 42)
    assert sm.read_float32(10)[0] == 42

    sm.write_string(10, "forty-two")
    assert sm.read_string(10)[0] == "forty-two"

    assert os.path.exists(sm._hook.file_path)
    sm.delete()
    assert not os.path.exists(sm._hook.file_path)
    
def test_timeout():
    delete_files("test")
    sm = SharedMemory("test", 100, True, 0.01)
    sm.write_bool(0, True)
    sm.give_control(wait = False)
    with pytest.raises(TimeoutError):
        sm.read_bool(0)

def test_two_python():
    delete_files("test")
    sm0 = SharedMemory("test", 100, True, 30)
    sm1 = SharedMemory("test", 100, False, 30)
    sm0.write_string(42, "foo")
    assert sm1.read_string(42)[0] == "foo"
    delete_files("test")

def test_resize():
    delete_files("test")
    sm0 = SharedMemory("test", 100, True, 0)
    sm1 = SharedMemory("test", 100, False, 0)
    sm0.write_string(42, "foo")
    assert sm1.read_string(42)[0] == "foo"
    sm1.resize(101)
    assert sm0.read_string(42)[0] == "foo"
    assert sm1.read_string(42)[0] == "foo"

    sm0.close()
    with pytest.raises(ConnectionError):
        sm1.read_bool(0)
    delete_files("test")

File: com.python-unity-shared-memory/Tests/Editor/Test_SM.cs
using NUnit.Framework;
using UnityEngine;
using System.IO;
using PythonUnitySharedMemory;

namespace PythonUnitySharedMemory.Tests.Editor{
   
    public class TestSharedMemory
    {

        [Test]
        public void TestCreate()
        {
            SharedMemory.DeleteFiles("test");
            var sm = new SharedMemory("test", 100, true, 30);
            sm.Close();

            var path = Path.Combine(Path.GetTempPath(), "python_unity_shared_memory/test");
            Assert.True(File.Exists(path));
            SharedMemory.DeleteFiles("test");
            Assert.False(File.Exists(path));
        }

        [Test]
        public void TestReadWrite(){
            SharedMemory.DeleteFiles("test");
            var sm = new SharedMemory("test", 100, true, 30);

            Assert.False(sm.ReadBool(0).Item1);
            sm.WriteBool(0, true);
            Assert.True(sm.ReadBool(0).Item1);

            Assert.AreEqual(sm.ReadInt32(3).Item1, 0);
            sm.WriteInt32(3, 42);
            Assert.AreEqual(sm.ReadInt32(3).Item1, 42);


            // Assert.AreApproximatelyEqual(sm.ReadFloat32(3).Item1, 0);
            // sm.WriteFloat32(3, 42);
            // comparer.AreApproximatelyEqual(sm.ReadFloat32(3).Item1, 42);

            sm.WriteString(10, "forty-two");
            Assert.AreEqual(sm.ReadString(10).Item1, "forty-two");

            var path = Path.Combine(Path.GetTempPath(), "python_unity_shared_memory/test");
            Assert.True(File.Exists(path));
            sm.Delete();
            Assert.False(File.Exists(path));
        }

        [Test]
        public void TestTimeout(){
            SharedMemory.DeleteFiles("test");
            var sm = new SharedMemory("test", 100, true, 0.1f);
            sm.WriteBool(0, true);
            sm.GiveControl(wait:false);
            Assert.Throws<System.TimeoutException>(() => sm.ReadBool(0));
        }

        [Test]
        public void TestTwoUnity(){
            SharedMemory.DeleteFiles("test");
            var sm0 = new SharedMemory("test", 100, true, 30);
            var sm1 = new SharedMemory("test", 100, false, 30);
            sm0.WriteString(42, "foo");
            Assert.AreEqual(sm1.ReadString(42).Item1, "foo");
            SharedMemory.DeleteFiles("test");
        }

        [Test]
        public void TestResize(){
            SharedMemory.DeleteFiles("test");
            var sm0 = new SharedMemory("test", 100, true, 30);
            var sm1 = new SharedMemory("test", 100, false, 30);
            sm0.WriteString(42, "foo");
            Assert.AreEqual(sm1.ReadString(42).Item1, "foo");
            sm1.Resize(101);
            Assert.AreEqual(sm0.ReadString(42).Item1, "foo");
            Assert.AreEqual(sm1.ReadString(42).Item1, "foo");
            sm0.Close();
            Assert.Throws<System.Exception>(() => sm1.ReadBool(0));
            SharedMemory.DeleteFiles("test");
        }
    }
}